### DISCUSSION TITLE
Listen to trajectory (modelIndex) frame change

### DISCUSSION THREAD

**Discussion by rjdirisio** [molstar/molstar#1261] [Q&A]
Hi, we use molstar to view trajectories. Outside of the trajectory viewer, we have an external component that we want to update when the frame (`modelIndex`) changes via the molstar `Next Model`, `Previous Model`, etc. buttons. What would be the best way to go about this?

**Comment by dsehnal** [comment]
`plugin.state.data` or the hierarchy manager have some events you can subscribe to. You need to watch for changes in the Model transform and communicate that to the rest of your app. You can look at the trajectory controls in mol-plugin-ui as that uses this logic.

**Reply by rjdirisio** [reply]
Thanks, I am able to get the model index from the label of the obj.label when I do something like
```ts
  plugin?.state.events.object.updated.subscribe((e) => {
    if (
      e.obj &&
      e.action === "recreate" &&
      e.obj.type === PSO.Molecule.Model.type
    ) {
      const modelLabel = e.obj.label;
      const newFrame = Number(modelLabel.split(" ")[1]);
    ...
    }
```
But, each time the button is pressed, I get ~100 events. Is there a way for me to be more surgical about only getting the 'new frame' event?

**Reply by dsehnal** [reply]
Yeah, you can do `plugin?.state.events.object.updated.pipe(filter(...), throttleTime(...), debounceTime(...), map(...), distinctUntilChanged(...), etc).subscribe(....)`.

See https://rxjs.dev/api/operators for options.

**Reply by dsehnal** [reply]
Also it might be better to subscribe to `this.plugin.state.data.events.changed` and you can use `this.plugin.behaviors.state.isAnimating` too determine if you actually want to trigger your update or not.

A good rxjs combinator is `combineLatest` to react when either property changes.
